### ISSUE TITLE
[Feature Request] Engine checkpointing

### ISSUE THREAD

**Issue by pkdogcom** [pytorch/ignite#62]
When training models, I usually would like to save a checkpoint of the model every few epochs or minutes/hours, so that I can resume training from the latest checkpoint when there are errors either in the codes or hardwares, or when I find it necessary to finetune the learning rate without losing the progress. 

In Tensorflow, there are some helper functions in the tf.train package to simplify checkpoint handling, like looking up for latest checkpoints As I was wondering if ignite can also include some of those features so that we don't need to write too many boilerplate codes.

**Comment by pkdogcom** [comment]
It is also very common to keep track of the best checkpoint and it will nice if ignite has out-of-box support for this feature.

**Comment by alykhantejani** [comment]
hi @pkdogcom,

Thanks for the feature request. I agree these both sound like nice features to have. I will add these to my current backlog and start working on them soon if no one else from the community pick them up.

**Comment by pkdogcom** [comment]
And I think it will be also very nice if this feature is implemented with logging/plotting integrated. For example, I've been using Ryzen machine for training recently and it is not stable on Linux and the whole system may hang and require rebooting every now and then. In such case it will be nice for the training to be able to resume from the latest checkpoint, not only with the model and optimizer parameters correctly restored, but also with the state of training related to plotting, such as iteration/epoch counts, Tensorboard events (probably need to remove events logged after the last checkpoint) recovered correctly, so that the training can continue as if it hasn't stopped at all.

**Comment by elanmart** [comment]
@pkdogcom what kind of API do you have in mind? In most of my workflows I simply use 

```python
@engine.on(Events.EPOCH_COMPLETED)
def checkpoint(engine):
   if engine.current_epoch % config.checkpoint_freq == 0:
        torch.save(model, config.storage / 'model.pth')
```

**Comment by jasonkriss** [comment]
@elanmart @alykhantejani I think this opens up a broader question... Do we prefer to keep this library as small as possible or do we prefer to remove as much boilerplate as possible. I think @elanmart's snippet is a great example of how well setup the current API is to make things like checkpointing very easy/explicit. That being said, there may be an argument for making this even easier since it is such a common use case. For example, something like...

```python
engine.add_event_handler(Events.EPOCH_COMPLETED, Checkpoint(config.storage, config.checkpoint_freq))
```

I personally don't have a strong opinion either way. The explicitness of @elanmart's snippet is attractive, but so is the conciseness of the built-in handler. It might be worth having a higher-level discussion around the goals of this project before implementing things like this.

**Comment by elanmart** [comment]
@jasonkriss 
> Do we prefer to keep this library as small as possible or do we prefer to remove as much boilerplate as possible.

I have quite a strong preference for keeping it as simple as possible. I just find it easier to read, and also enjoy having as much flexibility as possible.

That being said, maybe something like `contrib` module, that is build to match the "main" API would be nice? I would vote against changing the API itsetlf. 

But then again, I guess @alykhantejani 's opinion would be the most important here :)

**Comment by pkdogcom** [comment]
@elanmart I think your code snippet works for the ideal case, but in my experiences dealing with unstable system (say Ryzen) and large dataset like Imagenet or Kinetics which takes days or even weeks to train, it is also very crucial to be able to recover from the latest checkpoint in case that training fails accidentally, with all the training states like model weights, current learning rates, epochs, logging events recovered correctly as if training has never failed. And I think this can take some non-trivial amount of boilerplates code to implement (at least it is not a one-liner with the current API I guess) .

**Comment by jasonkriss** [comment]
@elanmart makes total sense to me. I agree on not wanting to change the main API too much. A `contrib` module might be a good idea.

**Comment by alykhantejani** [comment]
In my opinion ignite should achieve the following (as @jasonkriss once put it): "make the simple things simple and the hard things possible", therefore I would also prefer not to make the API too complex.

I believe that with the current event/callback model most of these things can be achieved with custom callbacks/handlers. These could live in `ignite.handlers` and we could also have an `ignite.contrib.handlers` which contains many more but maybe not as generic handlers?

However, that being said, I think a `engine.save()` function might be quite useful if it serializes:
1) the model
2) the optimizer
3) the state & history
4) allows easily resuming training from this state.

This is not an easy thing to get right, but then makes re-starting crashed training very easy, without loosing lots of work done. Thoughts?

**Comment by elanmart** [comment]
@alykhantejani how do you think the training should be resumed? Save every (`n`th) epoch and recover from that? Save every `n`th iteration? 

If you'd like to save every `n`th iteration, and let's say you'd like to resume training in the middle of an epoch, how would you go about it? Skip the initial batches? Finish that epoch early?

**Comment by alykhantejani** [comment]
@elanmart I think saving at iterations would be preferable as sometimes epoch boundaries can be very far apart (e.g. imagenet). Resuming from mid-epoch is a bit tricky, I guess we can choose to define it as:

If an epoch is 10 batches, and the engine is saved at batch 5, when resumed only 5 batches will be consumed for the remaining epoch. However, we can't guarantee that you wont see the same data twice in that epoch, as this would also require saving the dataloader/sampler etc (which actually may be possible).

**Comment by pkdogcom** [comment]
I think you may follow the DistributedSampler in core Pytorch as an example and define a new sampler, which use epoch as the deterministic manual seed and thus generate deterministic shuffled (subset) samples, which allow the dataloader to resume deterministically given the batch index. The sampler itself can be saved and manipulated by the engine depending on whether the current epoch is started from beginning (thus the sampler returns the full shuffled indices) or is resumed from the middle (thus the sampler only returns the subset of indices starting from the given batch index saved at checkpoint).

**Comment by alykhantejani** [comment]
Renamed the issue as model checkpointing has been added in #75

**Comment by pkdogcom** [comment]
@alykhantejani I've checked out the recent commits of checkpointing. It doesn't seem to have out-of-box support of resuming training from latest checkpoint. Is it under recent development plan?

**Comment by alykhantejani** [comment]
Hi @pkdogcom, you're correct - the current model checkpointing handler is a just for convenience of saving models, not saving the engine state. This is still something that I think would be good - let's keep this issue open to track it and it's definitely on the roadmap

**Comment by pkdogcom** [comment]
Sounds good. Thanks.

**Comment by vfdev-5** [comment]
Maybe we can inspire from [here](https://github.com/ncullen93/torchsample/blob/master/torchsample/callbacks.py)

**Comment by alykhantejani** [comment]
This is something I'll get started on once we ship 0.1.0 (should be soon)

**Comment by veugene** [comment]
I mentioned in #116 that it would be useful to be able to set the initial epoch in the trainer to allow resuming at some epoch. Would such an option be redundant if engine save/load is introduced? May there be cases where one would want to resume training at some non-zero epoch without loading the full engine state?

>In my opinion ignite should achieve the following (as @jasonkriss once put it): "make the simple things simple and the hard things possible", therefore I would also prefer not to make the API too complex.

Regarding the overall flexibility vs complexity discussion, I would expect a meta-framework (one that wraps around a framework such as pytorch or theano) to make a strong effort toward standardizing common code without restricting how I use the underlying framework.

The reasons I invest in the use of a meta-framework are (1) to standardize (and modularize) code so that it can be cleanly shared, (2) to assume a common workflow in a group project without much time wasted developing project-specific patterns, and (3) to reduce boilerplate code for my own convenience.

I think ignite is moving in the right direction : retain maximum flexibility while providing abstractions for common code. Additional features like `engine.save` achieve the latter nicely -- the added API complexity is not a problem as long as the API is not made more confusing. It's good for it to be feature-rich; we should just want to avoid anti-patterns.

> However, that being said, I think a engine.save() function might be quite useful if it serializes:
> 1. the model
> 2. the optimizer
> 3. the state & history
> 4. allows easily resuming training from this state.

That said, it may be tricky to serialize custom functions (and closures) in the engine or state. That doesn't have to supported but it would be necessary if one were to go the route of saving and loading models along with the engine rather than just model state.

**Comment by vfdev-5** [comment]
> May there be cases where one would want to resume training at some non-zero epoch without loading the full engine state?

Maybe, `engine.save` could have an option to load only some parts: model weights, start epoch, some history...

**Comment by vfdev-5** [comment]
@alykhantejani @jasonkriss @pkdogcom guys, could we make some precisions on the type of `data` acceptable in `Engine.run(data)`? I mean that, actually, the [documentation](https://github.com/pytorch/ignite/blob/master/ignite/engine/engine.py#L188) says `data` is an iterable. In the most of examples, I think, we just use pytorch DataLoader which is an iterable with `__len__`. 
My problem is that to resume from an iteration, I need to know how much iterations is one epoch and make something like `iteration % len(data)`. What do you think, should we change the doc and restrict `data` to an iterable with `__len__` or introduce an argument as `epoch_length` ?

**Comment by jasonkriss** [comment]
Enforcing that `data` has a `__len__` method seems preferable to me. Are there any common situations where this won't be possible though? Personally, in my projects, this would never be a problem.

**Comment by alykhantejani** [comment]
Hi - sorry haven't been able to get to this yet (am away on vacation until tomorrow). I'll take a look first thing tomorrow, but on first glance I think we should enforce `__len__` as this would conform with the pytorch dataset interface

**Comment by pkdogcom** [comment]
There seems to be several major issues including this one which have been stuck and lack of discussion for months. Is there any way to make some decisions and move forward while leaving refactoring later if necessary?
